src/libpmemobj/redo.c

```diff
@@ -322,6 +347,75 @@ redo_log_entry_val_create(struct redo_log *redo, size_t offset, uint64_t *dest,
 	return e;
 }
 
+/*
+ * redo_log_entry_buf_create -- atomically creates a buffer entry in the log
+ */
+struct redo_log_entry_buf *
+redo_log_entry_buf_create(struct redo_log *redo, size_t offset, uint64_t *dest,
+	const void *src, uint64_t size,
+	enum redo_operation_type type, const struct pmem_ops *p_ops)
+{
+	struct redo_log_entry_buf *e =
+		(struct redo_log_entry_buf *)(redo->data + offset);
+
+	/*
+	 * Depending on the size of the source buffer, we might need to perform
+	 * up to three separate copies:
+	 *	1. The first cacheline, 24b of metadata and 40b of data
+	 * If there's still data to be logged:
+	 *	2. The entire remainder of data data aligned down to cacheline,
+	 *	for example, if there's 150b left, this step will copy only
+	 *	128b.
+	 * Now, we are left with between 0 to 63 bytes. If nonzero:
+	 *	3. Create a stack allocated cacheline-sized buffer, fill in the
+	 *	remainder of the data, and copy the entire cacheline.
+	 *
+	 * This is done so that we avoid a cache-miss on misaligned writes.
+	 */
+
+	struct redo_log_entry_buf *b = alloca(CACHELINE_SIZE);
+	b->base.offset = (uint64_t)(dest) - (uint64_t)p_ops->base;
+	b->base.offset |= REDO_OPERATION(type);
+	b->size = size;
+
+	size_t ncopy = MIN(size,
+		CACHELINE_SIZE - sizeof(struct redo_log_entry_buf));
+	memcpy(b->data, src, ncopy);
+
+	size_t remaining_size = ncopy > size ? 0 : size - ncopy;
+
+	char *srcof = (char *)src + ncopy;
+	size_t rcopy = ALIGN_DOWN(remaining_size, CACHELINE_SIZE);
+	size_t lcopy = remaining_size - rcopy;
+
+	uint8_t last_cacheline[CACHELINE_SIZE];
+	if (lcopy != 0)
+		memcpy(last_cacheline, srcof + rcopy, lcopy);
```

Comment:
I think we should memset the remaining part of the cacheline, to be sure we don't depend on its previous state and don't leak some secret that happened to be on the stack put by user application.